File: version_java/src/main/java/com/example/version_java/data/source/remote/CriminalRemoteDataSourceImpl.java
package com.example.version_java.data.source.remote;

import androidx.annotation.NonNull;

import com.example.version_java.data.source.local.CrimianlLocalDataSource;
import com.example.version_java.network.api.SheetApi;
import com.example.version_java.network.response.CriminalResponse;

import java.util.List;

import javax.inject.Inject;

import kotlin.jvm.functions.Function1;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class CriminalRemoteDataSourceImpl implements CriminalRemoteDataSource {

    private final SheetApi sheetApi;


    @Override
    public void getRemoteCriminals(@NonNull Function1 onSuccess, @NonNull Function1 onFailure) {
        sheetApi.getSheetCriminals().enqueue(new Callback<List<CriminalResponse>>() {
            @Override
            public void onResponse(Call<List<CriminalResponse>> call, Response<List<CriminalResponse>> response) {
                if (response.body() != null) {
                    onSuccess.invoke(response.body());
                } else {
                    onFailure.invoke("GetRemoteCriminals Error");
                }
            }

            @Override
            public void onFailure(Call<List<CriminalResponse>> call, Throwable t) {
                onFailure.invoke(t.getMessage());
            }
        });
    }

    @Inject
    CriminalRemoteDataSourceImpl(
            SheetApi sheetApi
    ) {
        this.sheetApi = sheetApi;
    }
}

File: app/src/main/java/com/example/crimenotification/data/source/remote/CriminalRemoteDataSourceImpl.kt
package com.example.crimenotification.data.source.remote

import com.example.crimenotification.network.response.CriminalResponse
import com.example.crimenotification.network.api.SheetApi
import com.example.crimenotification.util.Result
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import javax.inject.Inject

class CriminalRemoteDataSourceImpl @Inject constructor(private val sheetApi: SheetApi) :
    CriminalRemoteDataSource {
    override suspend fun getCriminals(): Result<List<CriminalResponse>> = withContext(Dispatchers.IO) {
        try {
            val result = sheetApi.getSheetCriminals().execute().body()!!
            return@withContext Result.Success(result)
        } catch (e: Exception) {
            return@withContext Result.Error(e)
        }
    }
}